### PR TITLE
Handle more than 64 registers - Part 3

### PR DESCRIPTION
**Pull request by kunalspathak**
- ~This PR basically adds a `high` field in `regMaskTP`, but do not use it.~
EDIT: I will do this as a separate PR to isolate TP impact.
- Introduce a new type `SingleTypeRegSet` that denotes a 8-byte register set to represent registers from a single type i.e. gpr or float and in future predicated registers.
- Updated various places in LSRA that handles transition from `regMaskTP` and `SingleTypeRegSet`. Basically, all the fields in `RefPosition`, `Interval`, `RegisterSelection` are updated to be `SingleTypeRegSet` because they can never have mixed register set.

### CODE REVIEW

**Comment by kunalspathak** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=686947&view=ms.vss-build-web.run-extensions-tab)

image
image

**Review by EgorBo — APPROVED** [review]
Looks good to me from my limitted understanding

**Comment by kunalspathak** [comment]
@jakobbotsch - I am building things on top of this PR, so will go ahead and merge. Feel free to give a review once you are back. Thanks @EgorBo !

File: src/coreclr/jit/target.h [hunk 1]
```diff
@@ -371,6 +356,18 @@ static bool operator>(regMaskTP first, regMaskTP second)
     return first.getLow() > second.getLow();
 }
 
+static regMaskTP operator<<(regMaskTP& first, const int b)
+{
+    regMaskTP result(first.getLow() << b);
+    return result;
+}
+
+static regMaskTP operator>>(regMaskTP& first, const int b)
```
Comment (jakobbotsch) [inline src/coreclr/jit/target.h:365]: These shouldn't be taking `first` by (mutable) reference since they aren't the compound assignment versions. I'd suggest we just make it by value.

File: src/coreclr/jit/lsra.h [hunk 2]
```diff
@@ -2493,7 +2498,7 @@ class RefPosition
     // Prior to the allocation pass, registerAssignment captures the valid registers
     // for this RefPosition.
     // After the allocation pass, this contains the actual assignment
-    regMaskTP registerAssignment;
+    SingleTypeRegSet registerAssignment;
```
Comment (jakobbotsch) [inline src/coreclr/jit/lsra.h:2501]: This can contain multiple different register types for `RefTypeKill`. I think there are plenty of unused fields in `RefPosition` for `RefTypeKill` to fit a union that has a full `regMaskTP`.

Is it possible to double check that the other changed register sets in this PR don't also suffer from this problem? For example, some asserts somewhere to validate that `SingleTypeRegSet` don't have both float and int registers.

**Comment by MichalStrehovsky** [comment]
This change broke nativeaot outerloop runs on arm32: #102715.

File: src/coreclr/jit/target.h (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/target.h:365]: Yes, I changed bunch of these but forgot to change for `<<` and `>>`. Will fix in next PR.

File: src/coreclr/jit/lsra.h (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/lsra.h:2501]: > This can contain multiple different register types for RefTypeKill

I realized that while working on this PR, but wanted to have minimal changes to keep things unblocked. I will run into this problem anyway when I add predicate registers, because now, the `registerAssignment` for `RefTypeKill` need to technically hold more than 64 registers. One of the option as you suggested is to `union` some of the unused fields for `RefTypeKill` or convert `SingleTypeRegSet` to `64RegsSet` data type (same concept but different name to identify that former just had register set of one type vs. the newer one has 64 registers in a set of mixed type), but with the later approach, we might end up creating 2 `RefTypeKill` RefPosition - first for gpr/float and second for predicate. I will see which one is better approach and go with the minimalistic change needed at this point to unblock the predicate register implementation. 

> Is it possible to double check that the other changed register sets in this PR don't also suffer from this problem

There may be and the plan is to fix them up when I add predicate register, because then we will have to differentiate between which register type we are touching.

File: src/coreclr/jit/compiler.h [hunk 3]
```diff
@@ -11256,16 +11256,16 @@ class Compiler
     // This was done to avoid polluting all `targetXXX.h` macro definitions with a compiler parameter, where only
     // TARGET_AMD64 requires one.
     //
-    regMaskTP rbmAllFloat;
-    regMaskTP rbmFltCalleeTrash;
-    unsigned  cntCalleeTrashFloat;
+    SingleTypeRegSet rbmAllFloat;
+    SingleTypeRegSet rbmFltCalleeTrash;
+    unsigned         cntCalleeTrashFloat;
```
Comment (jakobbotsch) [inline src/coreclr/jit/compiler.h:11261]: Why these changes? I am not sure I'm sold that these cannot just stay as `regMaskTP`.

File: src/coreclr/jit/target.h [hunk 4]
```diff
@@ -742,15 +739,15 @@ inline bool floatRegCanHoldType(regNumber reg, var_types type)
 
 extern const regMaskSmall regMasks[REG_COUNT];
 
-inline regMaskTP genRegMask(regNumber reg)
+inline SingleTypeRegSet genRegMask(regNumber reg)
```
Comment (jakobbotsch) [inline src/coreclr/jit/target.h:742]: I am not sold on changing this from `regMaskTP` to `SingleTypeRegSet`. It is going to introduce ambiguity in the future about what register is being referred to.

As we discussed, I think we should isolate this handling to LSRA. LSRA can have its own context-dependent register sets where the first GP, float or predicate register all get mapped to 0, while the rest of the JIT can keep using the large register sets. LSRA can handle mapping these back to `regNumber` or `regMaskTP` (in the rare cases) to interact with the rest of the JIT.

For `instrDesc` we can stop storing `regMaskTP` in `instrDescCGCA` and just have an accessor method that maps the internal storage back to `regMaskTP`.
